### PR TITLE
add metric MergeTreeIndexGranularityInternalArraysTotalSize

### PR DESCRIPTION
**Pull request by mstetsyuk**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add the new MergeTreeIndexGranularityInternalArraysTotalSize metric to system.metrics. This metric is needed to find the instances with huge datasets susceptible to the high memory usage issue.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 7504e7692f8ec9dd2b76ca5249927604d166a792 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/72490/7504e7692f8ec9dd2b76ca5249927604d166a792/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/72490/7504e7692f8ec9dd2b76ca5249927604d166a792/style_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Common/CurrentMetrics.cpp [hunk 1]
```diff
@@ -354,6 +354,8 @@
     M(SharedCatalogDropZooKeeperThreadsScheduled, "Number of queued or active jobs in the threadpool for drop of object in ZooKeeper in Shared Catalog.") \
     \
     M(SharedDatabaseCatalogTablesInLocalDropDetachQueue, "Number of tables in the queue for local drop or detach in Shared Catalog.") \
+    \
+    M(MergeTreeIndexGranularityInternalArraysTotalSizeMetric, "The total size of all internal arrays in Merge Tree index granularity objects.") \
```
[the remark this answers is not in the document]
Comment (tavplubix) [reply src/Common/CurrentMetrics.cpp:358]: also the "Metric" suffix is redundant

File: src/Storages/MergeTree/MergeTreeIndexGranularityAdaptive.cpp [hunk 2]
```diff
@@ -60,6 +66,8 @@ void MergeTreeIndexGranularityAdaptive::appendMark(size_t rows_count)
     {
         marks_rows_partial_sums.push_back(marks_rows_partial_sums.back() + rows_count);
     }
+
+    CurrentMetrics::add(CurrentMetrics::MergeTreeIndexGranularityInternalArraysTotalSize, sizeof(rows_count));
```
Comment (tavplubix) [inline src/Storages/MergeTree/MergeTreeIndexGranularityAdaptive.cpp:70]: should we also subtract it when the part gets removed?

Comment (mstetsyuk) [reply src/Storages/MergeTree/MergeTreeIndexGranularityAdaptive.cpp]: fixed

**Review by alexey-milovidov — COMMENTED** [review]
Looks good.

Add a "smoke test" to show that the metric exists and shows a non-zero value.

What do you think about adding a column to `system.tables` and `system.parts`?

**Comment by alesapin** [comment]
FYI https://github.com/ClickHouse/ClickHouse/pull/71658

**Review by tavplubix — APPROVED** [review]

**Pull request merged**